axlearn/common/layers.py

```diff
@@ -814,8 +821,14 @@ def _create_layer_parameter_specs(self) -> dict[str, ParameterSpec]:
         )
 
     def forward(self, x: Tensor) -> Tensor:
+        cfg = self.config
```

Comment:
Same question here, can we handle this inside XLA compiler directly instead of making it explicit here? Adding this kind of fine-level sharding spec is a divergence from GSPMD programming paradigm imho.

Reply:
Responded above with a concrete example. The main issues we are trying to solve here are 

1. We want Norms/Dropouts/Residuals to be computed in sequence parallel so partitioned along sequence dimension. This is to avoid redundant compute across tensor parallel workers.
2.. OpenXLA ShardingPropagation often makes non optimal decisions - [https://github.com/openxla/xla/blob/main/xla/service/sharding_propagation.h]([https://github.com/openxla/xla/blob/main/xla/service/sharding_propagation.h])
3. Those non optimal decisions such as say propagating sequence level sharding to the left hand side of the attention QKV proj or MLP up proj will cause the SPMD partitioner to insert all-to-alls or collective-permutes or involuntary full rematerialization.


I am curious, why Axlearn has an activation level sharding spec in use today for `linear2.output_partition_spec`
Was this added to also address non optimal SPMD partitioning or propagation?